raft/server.cc

```diff
@@ -1267,20 +1259,30 @@ future<> server_impl::enter_joint_configuration(server_address_set c_new, seasta
     if (joining.size() == 0 && leaving.size() == 0) {
         co_return;
     }
+
     _stats.add_config++;
-    co_await add_entry_internal(raft::configuration{std::move(c_new)}, wait_type::committed, as);
-}
+    const auto& e = _fsm->add_entry(raft::configuration{std::move(c_new)});
 
-future<> server_impl::set_configuration(server_address_set c_new, seastar::abort_source* as) {
-    co_await enter_joint_configuration(std::move(c_new), as);
-    // Above we co_wait that the joint configuration is committed.
+    // We've just submitted a joint configuration to be committed.
     // Immediately, without yield, once the FSM discovers
-    // this, it appends non-joint entry. Hence,
-    // at this point in execution, non-joint entry is in the log.
-    // By waiting for a follow up dummy to get committed, we
-    // automatically wait for the non-joint entry to get
-    // committed.
-    co_await add_entry_internal(log_entry::dummy(), wait_type::committed, as);
+    // joint configuration, it appends a corresponding non-joint entry.
+    // By waiting for the joint configuration first we ensure
+    // that the next non-joint configuration we get from fsm in io_fiber
+    // would be the one corresponding to our joint configuration,
+    // no matter if the leader changed in the meantime.
+
+    auto f = _conf_commit_promise.emplace().get_future();
+    try {
+        co_await wait_for_entry({.term = e.term, .idx = e.idx}, wait_type::committed, as);
+    } catch (...) {
+        // We need to 'observe' possible exceptions in f, otherwise they will be
+        // considered unhandled and cause a warning.
+        (void)f.handle_exception([id = _id] (auto e) {
+            logger.trace("[{}] error while waiting for non-joint configuration to be committed: {}", id, e);
+        });
+        throw;
+    }
+    co_await std::move(f);
```

Comment:
Isn't it possible that both the conf_commit_promise and wait_for_entry resolve successfully and we're still in joint?

Reply:
Ah, now I see that the future is only resolved when the configuration is non-joint... Well, I guess it's best to call it _non_joint_conf_commit_future